Add Trino 431 release notes (#19497)
trinodb/trino: docs/src/main/sphinx/release/release-431.md

  with the `hive.metastore.thrift.client.connect-timeout` and
  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)

## Hudi connector

* Replace the Hive metastore configuration property `hive.metastore-timeout`
  with the `hive.metastore.thrift.client.connect-timeout` and
  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)

## Iceberg connector

* Replace the Hive metastore configuration property `hive.metastore-timeout`
  with the `hive.metastore.thrift.client.connect-timeout` and
  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)

## Ignite connector

## JMX connector

## Kafka connector

## Kinesis connector

## Kudu connector

## Local file connector

## MariaDB connector

## Memory connector

## MongoDB connector

## MySQL connector

## Oracle connector

## Phoenix connector

## Pinot connector

## PostgreSQL connector

## Prometheus connector

## Raptor connector

## Redis connector

## Redshift connector

## SingleStore connector

## SQL Server connector

## TPC-H connector

## SPI

* Add `ValueBlock` abstraction along with `VALUE_BLOCK_POSITION` and

>   `VALUE_BLOCK_POSITION_NOT_NULL` calling convention. ({issue}`19385`)

Reviewer: ```suggestion
  `VALUE_BLOCK_POSITION_NOT_NULL` calling conventions. ({issue}`19385`)
```

> * Require a separate block position for each argument for aggregation functions
>   for each argument. ({issue}`19385`)

Reviewer: ```suggestion
* Require a separate block position for each argument of aggregation functions. ({issue}`19385`)
```

# Release 431 (25 Oct 2023)

## General

> * Add support for `CREATE OR REPLACE TABLE` statement. ({issue}`13180`)

Reviewer 2: ```suggestion
* Add support for the `CREATE OR REPLACE TABLE` statement. ({issue}`13180`)
```

> * Disallow a null offset in {func}`lead` and {func}`lag`. ({issue}`19003`)

Reviewer 2: ```suggestion
* Disallow an offset of `null` in the {func}`lead` and {func}`lag` functions. ({issue}`19003`)
```
I think this is a little clearer.

## JDBC driver

> * Add a new JDBC property `legacyPreparedStatements` that allows to control

Reviewer 2: ```suggestion
* Add a new JDBC property `legacyPreparedStatements` that controls
```

  whether `EXECUTE IMMEDIATE` is used to execute prepared statements. ({issue}`18957`)

## Delta Lake connector

> * Replace the Hive metastore configuration property `hive.metastore-timeout`

Reviewer 2: ```suggestion
* Replace the `hive.metastore-timeout` Hive metastore configuration property
```
Re-order for consistency w/other notes and flow.
etc. for the other notes.

# Release 431 (25 Oct 2023)

## General

> * Add support for the `CREATE OR REPLACE TABLE` statement. ({issue}`13180`)

Reviewer 3: `CREATE TABLE` already exists. Technically, this adds `REPLACE TABLE`. It can only be used in combination with `CREATE TABLE`, in the form of `CREATE OR REPLACE TABLE`.

Author: I know .. how do we say that ... 

Add support for `OR REPLACE` with `CREATE TABLE` statements.  

??

> * Disallow a offset of `null` in the {func}`lead` and {func}`lag` functions. ({issue}`19003`)

Reviewer: `an`, but this sounds off to me.  Maybe:
```suggestion
* Disallow `null` value for the offset of the {func}`lead` and {func}`lag` functions. ({issue}`19003`)
```

Reviewer 3: Disallow a `null` offset ...

Author: Adjusted

## Web UI

* Mask shared secret value in logs. ({issue}`19519`)

## JDBC driver

> * Add a new JDBC property `legacyPreparedStatements` that controls whether

Reviewer 3: This is too low level. What's the user visible impact? (something like "improve query latency by reducing number of roundtrips when executing prepared statements")

Author: Yeah .. well.. it does do that since its off by default ..

  whether `EXECUTE IMMEDIATE` is used to execute prepared statements. ({issue}`18957`)

## Delta Lake connector

* Replace the Hive metastore configuration property `hive.metastore-timeout`
  with the `hive.metastore.thrift.client.connect-timeout` and
  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)

## Hive connector

* Replace the Hive metastore configuration property `hive.metastore-timeout`
  with the `hive.metastore.thrift.client.connect-timeout` and
  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)

> * Mask Hive S3 proxy password value in logs. ({issue}`19519`)

Author: Do these two have to go into Iceberg, Delta Lake and Hudi as well @wendigo ?

Reviewer: I'm pretty sure they are used there also... the file system code is shared across all of them

* Mask key store and trust store password values used for Hive metastore access

>   in logs. ({issue}`19519`)

Reviewer: same

* Add support for the `REPLACE` modifier to the `CREATE TABLE` statement. ({issue}`13180`)
* Disallow a `null` offset for the {func}`lead` and {func}`lag` functions. ({issue}`19003`)
* Improve performance of queries with short running splits. ({issue}`19487`)

## Security

* Support defining rules for procedures in file-based access control. ({issue}`19416`)
* Mask additional sensitive values in log files. ({issue}`19519`)

## JDBC driver

* Add a new JDBC property `legacyPreparedStatements` that controls whether
  `EXECUTE IMMEDIATE` is used to execute prepared statements. ({issue}`18957`)

## Delta Lake connector

* Replace the `hive.metastore-timeout` Hive metastore configuration property
  with the `hive.metastore.thrift.client.connect-timeout` and
  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)

## Hive connector

* Add support for [SQL routine management](sql-routine-management). ({issue}`19308`)
* Replace the `hive.metastore-timeout` Hive metastore configuration property
  with the `hive.metastore.thrift.client.connect-timeout` and
  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)

> * Improve support for concurrent table statistics updates in Glue. ({issue}`19463`)

Author: fyi @findepi and @findinpath .. I reworded to this. Also does this apply to Iceberg, Delta Lake, or Hudi connectors as well?

Reviewer 4: To my knowledge it applies only to Hive.

Author: Thanks

# Release 431 (27 Oct 2023)

## General

* Add support for [](/routines). ({issue}`19308`)
* Add support for [](/sql/create-function) and [](/sql/drop-function) statements. ({issue}`19308`)
* Add support for the `REPLACE` modifier to the `CREATE TABLE` statement. ({issue}`13180`)
* Disallow a `null` offset for the {func}`lead` and {func}`lag` functions. ({issue}`19003`)
* Improve performance of queries with short running splits. ({issue}`19487`)

## Security

* Support defining rules for procedures in file-based access control. ({issue}`19416`)
* Mask additional sensitive values in log files. ({issue}`19519`)

## JDBC driver

> * Allow configuration of executing statements with the legacy

Reviewer 3: This doesn't say what's the impact. I would rephrase as something along the lines of:

> Improve latency for prepared statements for Trino versions that support `EXECUTE IMMEDIATE` when the `explicitPrepare` parameter to is set to `false`.

Author: Updated

  `PREPARE...EXECUTE` or the faster `EXECUTE IMMEDIATE` invocation.
  ({issue}`19541`)

## Delta Lake connector

* Replace the `hive.metastore-timeout` Hive metastore configuration property
  with the `hive.metastore.thrift.client.connect-timeout` and
  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)

## Hive connector

* Add support for [SQL routine management](sql-routine-management). ({issue}`19308`)
* Replace the `hive.metastore-timeout` Hive metastore configuration property
  with the `hive.metastore.thrift.client.connect-timeout` and
  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)

> * Cast char fields, if necessary, to varchar type in Hive view translations. ({issue}`18337`)

Reviewer 3: That's the implementation detail. What problem is this solving?

Author: Prevents query translations failures.. updated

> * Improve support for concurrent table statistics updates in Glue. ({issue}`19463`)

Reviewer 3: "Improve support for concurrent updates of table statistics in Glue."
